src/test/groovy/pl/futurecollars/invoicing/controller/InvoiceControllerIntegrationTest.groovy

```diff
@@ -0,0 +1,189 @@
+package pl.futurecollars.invoicing.controller
+
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.http.MediaType
+import org.springframework.test.web.servlet.MockMvc
+import org.springframework.test.web.servlet.ResultActions
+import pl.futurecollars.invoicing.helpers.TestHelpers
+import pl.futurecollars.invoicing.model.Invoice
+import pl.futurecollars.invoicing.util.JsonService
+import spock.lang.Specification
+
+import java.time.LocalDate
+
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
+
+@SpringBootTest
+@AutoConfigureMockMvc
+class InvoiceControllerIntegrationTest extends Specification {
+
+    @Autowired
+    MockMvc mockMvc
+    @Autowired
+    JsonService jsonService
+    Invoice invoice = TestHelpers.invoice(1)
+    private LocalDate updatedDate = LocalDate.of(2021, 05, 03)
+
+    def setup() {
+        getAllInvoices().each { invoice -> deleteInvoice(invoice.id)}
+    }
+
+    def "empty array is returned when no invoices were added"() {
+        expect:
+        getAllInvoices() == []
+    }
+
+    def "returned correctly id when invoice is added"() {
+        given:
+        def invoiceAsJson = convertToJson()
+
+        expect:
+//        addInvoice(invoiceAsJson) == 1
+//        addInvoice(invoiceAsJson) == 2
+//        addInvoice(invoiceAsJson) == 3
+        def id = addInvoice(invoiceAsJson)
+        addInvoice(invoiceAsJson) == id + 1
+        addInvoice(invoiceAsJson) == id + 2
+        addInvoice(invoiceAsJson) == id + 3
+
+    }
+
+    def "returned all invoices when getting all invoices"() {
+        given:
+        def count = 3
+        def expectedInvoices = addUniqueInvoices(count)
+
+        expect:
+        getAllInvoices().size() == count
+        getAllInvoices() == expectedInvoices
+    }
+
+    def "returned invoice when getting by id"() {
+
+        given:
+        def expectedInvoice = addUniqueInvoices(5)
+        def verifiedInvoice = expectedInvoice.get(2)
+        def id = verifiedInvoice.getId()
+
+        when:
+        def response = mockMvc.perform(get("/invoices/$id"))
```

Comment:
wyciągnął bym tutaj "/invoices" do stałej, bo powtarza się w kodzie wiele razy - gdyby trzeba było zmienić nazwę albo dodać nr wersji API ( "/v1/invoices") to wtedy mamy tylko jedno miejsce do zmiany